### PR TITLE
Provide current_py_toolchain

### PR DESCRIPTION
**Pull request by jheaff1**
This PR introduces the current_py_toolchain rule, exposing the
PYTHON2 and PYTHON3 "make" variables in bazel rules, analagous to
@bazel_tools//tools/cpp:current_cc_toolchain.

This is useful when building 3rd party libraries which require python via rules_foreign_cc.

See
https://docs.bazel.build/versions/main/be/make-variables.html#custom_variables

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
Bazel rules can refer to the python interpreter via "Make" Variables. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by jheaff1** [comment]
CI now passes, may I please ask for a review?

File: python/defs.bzl [hunk 1]
```diff
@@ -41,6 +41,45 @@ def _add_tags(attrs):
...
 
+def _current_py_toolchain_impl(ctx):
+    toolchain = ctx.toolchains[ctx.attr._toolchain]
+
+    files = []
+    vars = {}
+
+    if toolchain.py3_runtime and toolchain.py3_runtime.interpreter:
+        files += [toolchain.py3_runtime.interpreter] + toolchain.py3_runtime.files.to_list()
+        vars = dict(vars.items() + {"PYTHON3": toolchain.py3_runtime.interpreter.path}.items())
```
Comment (f0rmiga) [inline python/defs.bzl:52]: Why not just:
```suggestion
        vars["PYTHON3"] = toolchain.py3_runtime.interpreter.path
```

File: python/defs.bzl [hunk 2]
```diff
@@ -41,6 +41,45 @@ def _add_tags(attrs):
...
+        vars = dict(vars.items() + {"PYTHON2": toolchain.py2_runtime.interpreter.path}.items())
+
+    return [
+        toolchain,
+        platform_common.TemplateVariableInfo(vars),
+        DefaultInfo(
+            runfiles = ctx.runfiles(files = files),
+            files = depset(files),
+        ),
+    ]
```
Comment (f0rmiga) [inline python/defs.bzl:65]: Using the depset and runfiles appropriate constructors:

```suggestion
    direct = []
    transitive = []
    vars = {}

    if toolchain.py3_runtime and toolchain.py3_runtime.interpreter:
        direct.append(toolchain.py3_runtime.interpreter)
        transitive.append(toolchain.py3_runtime.files)
        vars["PYTHON3"] = toolchain.py3_runtime.interpreter.path

    if toolchain.py2_runtime and toolchain.py2_runtime.interpreter:
        direct.append(toolchain.py2_runtime.interpreter)
        transitive.append(toolchain.py2_runtime.files)
        vars["PYTHON2"] = toolchain.py2_runtime.interpreter.path

    files = depset(direct, transitive = transitive)
    return [
        toolchain,
        platform_common.TemplateVariableInfo(vars),
        DefaultInfo(
            runfiles = ctx.runfiles(transitive_files = files),
            files = files,
        ),
    ]
```

**Review by f0rmiga — CHANGES_REQUESTED** [review]

File: python/defs.bzl [hunk 3]
```diff
@@ -41,6 +41,45 @@ def _add_tags(attrs):
...
+    other rules, such as genrule. It allows exposing a python toolchain after toolchain resolution has
+    happened, to a rule which expects a concrete implementation of a toolchain, rather than a
+    toolchain_type which could be resolved to that toolchain.
+    """,
+    implementation = _current_py_toolchain_impl,
+    attrs = {
+        "_toolchain": attr.string(default = str(Label("@bazel_tools//tools/python:toolchain_type"))),
+    },
+    toolchains = [
+        str(Label("@bazel_tools//tools/python:toolchain_type")),
```
Comment (f0rmiga) [inline python/defs.bzl:79]: Every time I see this, I wish we had a const for the toolchain type in the same way we have https://github.com/bazelbuild/bazel/blob/49610123de3c555f1f29dd4c6d91101c659a26b1/tools/cpp/toolchain_utils.bzl#L20.

**Comment by jheaff1** [comment]
@f0rmiga Thanks for the review, I have applied your suggested changes

File: python/defs.bzl (hunk 3 above)
Comment (jheaff1) [reply python/defs.bzl:83]: I can add such a constant changes change all references to `"@bazel_tools//tools/python:toolchain_type"` if you like, although that should probably be in a separate PR

Comment (f0rmiga) [reply python/defs.bzl:83]: Up to you if you want to contribute further.

**Comment by f0rmiga** [comment]
@jheaff1 thanks for addressing my comments!

**Review by f0rmiga — APPROVED** [review]

File: tests/load_from_macro/BUILD [hunk 4]
```diff
@@ -24,3 +24,11 @@ py_library(
     # Allow a test to verify an "outside package" doesn't get included
     visibility = ["//examples/wheel:__pkg__"],
 )
+
+genrule(
+    name = "test_current_py_toolchain",
+    srcs = [],
+    outs = ["out.txt"],
+    cmd = "$(PYTHON3) --version > $(location out.txt)",
```
Comment (aryeh-looker) [inline tests/load_from_macro/BUILD:32]: Is there a way to use `genrule` and a python toolchain in conjunction with pip deps?

Comment (aryeh-looker) [reply tests/load_from_macro/BUILD:32]: cc/ @ f0rmiga @jheaff1

Comment (jheaff1) [reply tests/load_from_macro/BUILD:32]: See the use of `entry_point` in [pip.md](https://github.com/bazelbuild/rules_python/blob/main/docs/pip.md#pip_parse)

you could use this in a genrule

Comment (jheaff1) [reply tests/load_from_macro/BUILD:32]: I’m assuming you want to run T he entrypoint of a pip package? If you want to use a pip package as a library, use the `requirement` macro and setup the PYTHONPATH environment variable to include the location of the pip packages. Eg

`PYTHONPATH=$(location {})`.format(requirement(‘the_pip_package`)

**Pull request merged**